dev/PagerControl/PagerControl.cpp

```diff
@@ -0,0 +1,776 @@
+
+static constexpr auto c_numberPanelButtonStyleName = L"NumberPanelButtonStyle"sv; 
+const int c_AutoDisplayModeNumberOfPagesThreshold = 10; // This integer determines when to switch between NumberBoxDisplayMode and ComboBoxDisplayMode 
+const int c_infiniteModeComboBoxItemsIncrement = 100;
+
+/* Common functions */
+PagerControl::PagerControl()
+{
+    __RP_Marker_ClassById(RuntimeProfiler::ProfId_PagerControl);
+
+    m_comboBoxEntries = winrt::make<Vector<IInspectable>>().as<winrt::IObservableVector<IInspectable>>();
+    m_numberPanelElements = winrt::make<Vector<IInspectable>>().as<winrt::IObservableVector<IInspectable>>();
+
+    const auto templateSettings = winrt::make<PagerControlTemplateSettings>();
+    templateSettings.SetValue(PagerControlTemplateSettingsProperties::s_PagesProperty, m_comboBoxEntries);
+    templateSettings.SetValue(PagerControlTemplateSettingsProperties::s_NumberPanelItemsProperty, m_numberPanelElements);
+    SetValue(s_TemplateSettingsProperty, templateSettings);
+
+    SetDefaultStyleKey(this);
+}
+
+PagerControl::~PagerControl()
+{
+    m_rootGridKeyDownRevoker.revoke();
+    m_comboBoxSelectionChangedRevoker.revoke();
+    m_firstPageButtonClickRevoker.revoke();
+    m_previousPageButtonClickRevoker.revoke();
+    m_nextPageButtonClickRevoker.revoke();
+    m_lastPageButtonClickRevoker.revoke();
+}
+
+void PagerControl::OnApplyTemplate()
+{
+    if (PrefixText().empty())
+    {
+        PrefixText(ResourceAccessor::GetLocalizedStringResource(SR_PagerControlPrefixTextName));
+    }
+    if (SuffixText().empty())
+    {
+        SuffixText(ResourceAccessor::GetLocalizedStringResource(SR_PagerControlSuffixTextName));
+    }
+    if (const auto rootGrid = GetTemplateChildT<winrt::FrameworkElement>(c_rootGridName, *this))
+    {
+        m_rootGridKeyDownRevoker = rootGrid.KeyDown(winrt::auto_revoke, { this, &PagerControl::OnRootGridKeyDown });
+    }
+
+    if (const auto firstPageButton = GetTemplateChildT<winrt::Button>(c_firstPageButtonName, *this))
+    {
+        m_firstPageButtonClickRevoker = firstPageButton.Click(winrt::auto_revoke, { this, &PagerControl::FirstButtonClicked });
+    }
+    if (const auto previousPageButton = GetTemplateChildT<winrt::Button>(c_previousPageButtonName, *this))
+    {
+        m_previousPageButtonClickRevoker = previousPageButton.Click(winrt::auto_revoke, { this, &PagerControl::PreviousButtonClicked });
+    }
+    if (const auto nextPageButton = GetTemplateChildT<winrt::Button>(c_nextPageButtonName, *this))
+    {
+        m_nextPageButtonClickRevoker = nextPageButton.Click(winrt::auto_revoke, { this, &PagerControl::NextButtonClicked });
+    }
+    if (const auto lastPageButton = GetTemplateChildT<winrt::Button>(c_lastPageButtonName, *this))
+    {
+        m_lastPageButtonClickRevoker = lastPageButton.Click(winrt::auto_revoke, { this, &PagerControl::LastButtonClicked });
+    }
+
+    if (const auto comboBox = GetTemplateChildT<winrt::ComboBox>(c_comboBoxName, *this))
```

Comment:
>GetTemplateChildT [](start = 30, length = 17)

With this method, if a new template is applied after the initial template, which removed this named template part, we would keep the pointer to the old template's part which would likely crash.  I would use a lambda here to limit the scope of the local variable while assigning the value of get template child to m_combobox event if the return value is null.

Reply:
Very good point, updated that section now.

Reply:
The method you chose to do this requires an extra call to m_combobox.get(), which isn't free, which is why i suggested a lambda (since it also scope limits.  However this does function.

---
In reply to: [494660076](https://github.com/microsoft/microsoft-ui-xaml/pull/3327#discussion_r494660076) [](ancestors = 494660076)

Reply:
I'm using lamdas now for the combobox and numberbox.